cmd/server/app/history_purge.go

```diff
@@ -0,0 +1,233 @@
+	Long:  `deletes all evaluation history entries older than 30 days, maintaining the latest one per rule/entity pair`,
+	RunE:  historyPurgeCommand,
+}
+
+func historyPurgeCommand(cmd *cobra.Command, _ []string) error {
+	batchSize := viper.GetUint("batch-size")
+	dryRun := viper.GetBool("dry-run")
+
+	cfg, err := config.ReadConfigFromViper[serverconfig.Config](viper.GetViper())
+	if err != nil {
+		cliErrorf(cmd, "unable to read config: %s", err)
+	}
+
+	ctx := logger.FromFlags(cfg.LoggingConfig).WithContext(context.Background())
+
+	// instantiate `db.Store` so we can run queries
+	store, closer, err := wireUpDB(ctx, cfg)
+	if err != nil {
+		cliErrorf(cmd, "unable to connect to database: %s", err)
+	}
+	defer closer()
+
+	threshold := time.Now().UTC().AddDate(-30, 0, 0)
+	fmt.Printf("Calculated threshold is %s", threshold)
+
+	if err := purgeLoop(ctx, store, threshold, batchSize, dryRun, cmd.Printf); err != nil {
+		cliErrorf(cmd, "failed purging evaluation log: %s", err)
+	}
+
+	return nil
+}
+
+// purgeLoop routine cleans up the evaluation history log by deleting
+// all records older than a fixed threshold.
+//
+// As of the time of this writing, the size of the row structs is 80
+// bytes, specifically
+//
+// * go time is 24 bytes
+// * rule id, entity id and evaluation id are 16 bytes UUIDs, adding
+// 48 bytes
+// * entity type, which is necessary because there's no guarantee that
+// the entity id is unique across entity types, is mapped to an
+// integer and adds another 16 bytes
+//
+// Given their size, 4 million records would allocate around 300MB of
+// RAM, adding some overhead for the used data structures we estimate
+// 500MB total RAM consumption in the worst case.
+//
+// From the execution time perspective, this is not necessarily the
+// best approach, and the time to first byte might become significant
+// as usage increases.
+func purgeLoop(
+	ctx context.Context,
+	store db.Store,
+	threshold time.Time,
+	batchSize uint,
+	dryRun bool,
+	printf func(format string, a ...any),
+) error {
+	total := 0
+	deleted := 0
+
+	records, err := store.ListEvaluationHistoryOlderThan(
+		ctx,
+		db.ListEvaluationHistoryOlderThanParams{
+			Threshold: threshold,
+			Size:      int32(4000000),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("error purging evaluation history: %w", err)
+	}
+
+	if len(records) == 0 {
+		printf("No records to delete\n")
+		return nil
+	}
+
+	total = len(records)
+	deletes := filterRecords(records)
+
+	if len(deletes) == 0 {
+		printf("No records to delete after filtering\n")
+		return nil
+	}
+
+	// Skip deletion if --dry-run was passed.
+	if !dryRun {
+		deleted, err = deleteEvaluationHistory(
+			ctx,
+			store,
+			deletes,
+			batchSize,
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	printf("Done purging history, deleted %d out of %d total records\n",
+		deleted,
+		total,
+	)
+
+	return nil
+}
+
+// filterRecords sift through the records separating the latest for
```

Comment:
I wonder if this could be implemented in the SQL query by using the `latest_evaluation_statuses` table and the `EXCEPT` operator.

Reply:
You're right, I integrated this suggestion, but I haven't used the `EXCEPT` operator and implemented it using joins.